test-data/unit/check-typeddict.test

```diff
@@ -3236,3 +3236,56 @@ def foo(x: int) -> Foo: ...
 f: Foo = {**foo("no")}  # E: Argument 1 to "foo" has incompatible type "str"; expected "int"
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-typeddict.pyi]
+
+
+[case testTypedDictWith__or__method]
+from mypy_extensions import TypedDict
+
+class Foo(TypedDict):
+    key: int
+
+foo1: Foo = {'key': 1}
+foo2: Foo = {'key': 2}
+
+reveal_type(foo1 | foo2)  # N: Revealed type is "TypedDict('__main__.Foo', {'key': builtins.int})"
+reveal_type(foo1 | {'key': 1})  # N: Revealed type is "TypedDict('__main__.Foo', {'key': builtins.int})"
+reveal_type(foo1 | {'key': 'a'})  # N: Revealed type is "typing.Mapping[builtins.str, builtins.object]"
+reveal_type(foo1 | {})  # N: Revealed type is "typing.Mapping[builtins.str, builtins.object]"
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+[case testTypedDictWith__or__method_error]
+from mypy_extensions import TypedDict
+
+class Foo(TypedDict):
+    key: int
+
+foo: Foo = {'key': 1}
+
+foo | 1
+[out]
+main:8: error: No overload variant of "__or__" of "TypedDict" matches argument type "int"
+main:8: note: Possible overload variants:
+main:8: note:     def __or__(self, Foo, /) -> Foo
+main:8: note:     def __or__(self, Mapping[str, object], /) -> Mapping[str, object]
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+# TODO: add `__ror__` method check, after `__ror__` definition is fixed.
+
+[case testTypedDictWith__ior__method]
+from mypy_extensions import TypedDict
+
+class Foo(TypedDict):
+    key: int
+
+foo: Foo = {'key': 1}
+foo |= {'key': 2}
+
+foo |= {}  # E: Missing key "key" for TypedDict "Foo"
+foo |= {'key': 'a', 'b': 'a'}  # E: Extra key "b" for TypedDict "Foo" \
+                               # E: Incompatible types (expression has type "str", TypedDict item "key" has type "int")
+foo |= {'b': 2}  # E: Missing key "key" for TypedDict "Foo" \
+                 # E: Extra key "b" for TypedDict "Foo"
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
```

Comment:
This test doesn't pass with the `master` branch checked out. But errors are emitted by mypy on exactly the same lines; the only thing that this PR seems to be changing is the specific wording of the error messages.

Reply:
Not exactly, if you add 

```python
[case testTypedDictWith__ior__method]
from mypy_extensions import TypedDict

class Foo(TypedDict):
    key: int

foo: Foo = {'key': 1}
foo |= {'key': 2}  # line 7

foo |= {}  # E: Missing key "key" for TypedDict "Foo"
foo |= {'key': 'a', 'b': 'a'}  # E: Extra key "b" for TypedDict "Foo" \
                               # E: Incompatible types (expression has type "str", TypedDict item "key" has type "int")
foo |= {'b': 2}  # E: Missing key "key" for TypedDict "Foo" \
                 # E: Extra key "b" for TypedDict "Foo"
[builtins fixtures/dict.pyi]
[typing fixtures/typing-typeddict.pyi]
```

to `master` and run it, you will get:

```
  main:7: error: Unsupported left operand type for | ("Foo") (diff)
  main:9: error: Unsupported left operand type for | ("Foo") (diff)
  main:10: error: Unsupported left operand type for | ("Foo") (diff)
  main:11: error: Unsupported left operand type for | ("Foo") (diff)
```

which is not correct, because code on line 7 must pass.

Reply:
If you _also_ apply the changes you've made to the `typing-typeddict.pyi` fixture, then you don't get the error on line 7 using `master`.

Reply:
hm, yes. this happens indeed. I think that I need to improve the test (this happens because I use `Mapping` in the test fixture). Thanks! But, here's the real result before and after using mypy-cli:

```python
from typing import TypedDict

class Foo(TypedDict):
    key: int

foo: Foo = {'key': 1}
foo |= {'key': 1}
reveal_type(foo)
foo |= {'key': 'a'}
```

Before:

```
ex.py:7: error: No overload variant of "__ror__" of "dict" matches argument type "Foo"  [operator]
ex.py:7: note: Possible overload variants:
ex.py:7: note:     def __ror__(self, dict[str, int], /) -> dict[str, int]
ex.py:7: note:     def [_T1, _T2] __ror__(self, dict[_T1, _T2], /) -> dict[str | _T1, int | _T2]
ex.py:8: note: Revealed type is "TypedDict('ex.Foo', {'key': builtins.int})"
ex.py:9: error: No overload variant of "__ror__" of "dict" matches argument type "Foo"  [operator]
ex.py:9: note: Possible overload variants:
ex.py:9: note:     def __ror__(self, dict[str, str], /) -> dict[str, str]
ex.py:9: note:     def [_T1, _T2] __ror__(self, dict[_T1, _T2], /) -> dict[str | _T1, str | _T2]
```

After:

```
ex.py:8: note: Revealed type is "TypedDict('ex.Foo', {'key': builtins.int})"
ex.py:9: error: Incompatible types (expression has type "str", TypedDict item "key" has type "int")  [typeddict-item]
```

I think that I might have an idea on how to make tests better and probably even add `__ror__` support.